iio/iio_app/iio_demo_app.c

```diff
@@ -0,0 +1,337 @@
+ */
+static ssize_t iio_demo_read_dev(void *iio_inst, char *pbuf, size_t offset,
+				 size_t bytes_count, uint32_t ch_mask)
+{
+	struct iio_demo_device *demo_device;
+	uint32_t i, j = 0, current_ch = 0;
+	uint16_t *pbuf16;
+	size_t samples;
+
+	demo_device = (struct iio_demo_device *)iio_inst;
+	pbuf16 = (uint16_t*)pbuf;
+	samples = (bytes_count * demo_device->num_channels) / hweight8(
+			  ch_mask);
+	samples /= 2; /* because of uint16_t *pbuf16 = (uint16_t*)pbuf; */
+	offset = (offset * demo_device->num_channels) / hweight8(ch_mask);
+
+	for (i = 0; i < samples; i++) {
+
+		if (ch_mask & BIT(current_ch)) {
+			pbuf16[j] = i;
+			j++;
+		}
+
+		if (current_ch + 1 < demo_device->num_channels)
+			current_ch++;
+		else
+			current_ch = 0;
+	}
+
+	return bytes_count;
+}
+
+/**
+ * @brief Get xml corresponding to device.
+ * @param xml - Xml containing description of a device.
+ * @param iio_dev - Structure describing a device, channels and attributes.
+ * @return SUCCESS in case of success or negative value otherwise.
+ */
+static ssize_t iio_demo_get_xml(char** xml, struct iio_device *iio_dev)
+{
+
+	*xml = calloc(1, strlen(demo_xml) + 1);
+	if (!(*xml))
+		return -ENOMEM;
+
+	memcpy(*xml, demo_xml, strlen(demo_xml));
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Application for reading/writing and parameterization of
+ * demo app.
+ * @param  desc - Application descriptor.
+ * @param init - Application configuration structure.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t iio_demo_app_init(struct iio_demo_app_desc **desc,
+			  struct iio_demo_app_init_param *init)
+{
+	struct iio_interface_init_par iio_demo_intf_par;
+	struct iio_device * iio_device;
+	int32_t status;
+	struct iio_demo_device *iio_demo_device_inst = (struct iio_demo_device *)calloc(
+				1,
+				sizeof(struct iio_demo_device));
+	if (!iio_demo_device_inst)
+		return FAILURE;
+
+	iio_demo_device_inst->name = demo_dev_name;
+	iio_demo_device_inst->num_channels = 8;
+	iio_device = iio_demo_create_device(demo_dev_name, 8);
+	if (iio_device) {
+		free(iio_demo_device_inst);
+		return FAILURE;
+	}
+
+	iio_demo_intf_par = (struct iio_interface_init_par) {
+		.dev_name = demo_dev_name,
+		.dev_instance = iio_demo_device_inst,
+		.iio_device = iio_device,
+		.get_xml = iio_demo_get_xml,
+		.transfer_dev_to_mem = iio_demo_transfer_dev_to_mem,
+		.read_data = iio_demo_read_dev,
+		.transfer_mem_to_dev = NULL,
+		.write_data = NULL,
+	};
+	status = iio_register(&iio_demo_intf_par);
+
+	if (status < 0) {
+		free(iio_demo_device_inst);
+		iio_demo_delete_device(iio_device);
+		return FAILURE;
+	}
+
+	return SUCCESS;
+}
+
+/**
+ * @brief Release resources.
+ * @param desc - Application descriptor.
+ * @return SUCCESS in case of success, FAILURE otherwise.
+ */
+int32_t iio_demo_app_remove(struct iio_demo_app_desc *desc)
+{
+	int32_t status;
+
+	if (!desc)
+		return FAILURE;
+
+	status = iio_unregister(desc->iio_demo_dev_inst->name);
+	if(status < 0)
+		return status;
```

Comment:
`iio_unregister` return values are `SUCCESS/FAILURE`, therefore you can return directly `FAILURE` here.